jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,429 @@
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.eclipse.jetty.util.component.Dumpable;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A fast container of poolable objects, with optional support for
+ * multiplexing and thread-local caching.
+ * @param <T>
+ */
+public class Pool<T> implements AutoCloseable, Dumpable
+{
+    private static final Logger LOGGER = Log.getLogger(Pool.class);
+
+    private final List<Entry> sharedList = new CopyOnWriteArrayList<>();
+    private final ThreadLocal<List<Entry>> cache;
+    private final Lock closeLock = new ReentrantLock();
+    private final int cacheSize;
+    private volatile boolean closed;
+    private volatile int maxMultiplex = 1;
+    private volatile int maxUsageCount = -1;
+
+    /**
+     * Construct a Pool with the specified thread-local cache size.
+     *
+     * @param cacheSize the thread-local cache size. A value less than 1 means the cache is disabled.
+     */
+    public Pool(int cacheSize)
+    {
+        this.cacheSize = cacheSize;
+        if (cacheSize > 0)
+            this.cache = ThreadLocal.withInitial(() -> (List<Entry>)new ArrayList<Entry>(cacheSize));
+        else
+            this.cache = null;
+    }
+
+    public int getMaxMultiplex()
+    {
+        return maxMultiplex;
+    }
+
+    public final void setMaxMultiplex(int maxMultiplex)
+    {
+        if (maxMultiplex < 1)
+            throw new IllegalArgumentException("Max multiplex must be >= 1");
+        this.maxMultiplex = maxMultiplex;
+    }
+
+    public int getMaxUsageCount()
+    {
+        return maxUsageCount;
+    }
+
+    public final void setMaxUsageCount(int maxUsageCount)
+    {
+        if (maxUsageCount == 0)
+            throw new IllegalArgumentException("Max usage count must be != 0");
+        this.maxUsageCount = maxUsageCount;
+    }
+
+    /**
+     * Create a new disabled slot into the pool. The returned entry
+     * won't be acquirable as long as {@link Entry#enable(Object)}
+     * has not been called.
+     *
+     * @return a disabled entry that is contained in the pool,
+     * or null if the pool is closed.
+     */
+    public Entry addReservation()
+    {
+        closeLock.lock();
+        try
+        {
+            if (!closed)
```

Comment:
I think we need to check max size within this lock, otherwise we will be in a test and set race.

Reply:
This makes sense when combined with the comment below.